tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c

```diff
@@ -344,16 +344,19 @@ static void alt_thread_entry(void)
 		"push {r0, r1};\n\t"
 		"mov r1, r7;\n\t"
 		"mov r0, %0;\n\t"
-		"ldmia r0!, {r4-r7};\n\t"
-		"ldmia r0!, {r4-r7};\n\t"
+		"ldr r4, [r0, #16];\n\t"
```

Comment:
How about keeping the 2 `ldmia` instructions instead of mixing `ldr` and `ldmia` here, like:

```
        "mov r0, %0;"
        "add r0, #16;"
        "ldmia r0!, {r4-r7};"
        "mov r8, r4;"
        "mov r9, r5;"
        "mov r10, r6;"
        "mov r11, r7;"
        "mov r0, %0;"
        "ldmia r0!, {r4-r7};"
```
? 

I think `load_callee_saved_regs` in line 64 should be fixed likewise -- for properly loading `r4-r7`?

Reply:
Thanks for the tips. Debugging, this causes a hardfault, though, because of:

```c
       "mov r0, %0;"
        "ldmia r0!, {r4-r7};"
```

GCC is resolving `%0` to `r7`, which we clobber in the first `"ldmia r0!, {r4-r7};"` (which sets `r7` to `0`), so the second `ldmia` call listed above is trying to increment `r0 = 0`, the same problem we had in the original issue.

I'll have to do some further debugging in the morning.

Reply:
Also ... adding `r7` to the clobber list causes a compilation error and CI failure as well, in some situations, since this register has special meaning in the [Thumb Procedure Call Standard](https://developer.arm.com/documentation/dui0041/c/Thumb-Procedure-Call-Standard/TPCS-definition/TPCS-register-names?lang=en).

Reply:
@gromero This should work. Rather than copying `%0` a second time (which has been clobbered as `r6` or `r7`), just decrement the pointer in `r0` by `#32`:

```
"mov r1, r7;\n\t"
"mov r0, %0;\n\t"
"add r0, #16;\n\t"
"ldmia r0!, {r4-r7};\n\t"
"mov r8, r4;\n\t"
"mov r9, r5;\n\t"
"mov r10, r6;\n\t"
"mov r11, r7;\n\t"
"sub r0, #32;\n\t"
"ldmia r0!, {r4-r7};\n\t"
```